mono/metadata/marshal.c

```diff
@@ -5935,9 +5943,17 @@ mono_marshal_type_size (MonoType *type, MonoMarshalSpec *mspec, guint32 *align,
 		klass = mono_class_from_mono_type_internal (type);
 		if (klass == mono_defaults.object_class &&
 			(mspec && mspec->native == MONO_NATIVE_STRUCT)) {
-		*align = 16;
-		return 16;
+			*align = 16;
+			return 16;
+		} 
+#if ENABLE_NETCORE
+		else if (strcmp (m_class_get_name_space (klass), "System") == 0 && 
```

Comment:
what does assume this alignment? wouldn't be better to use .Pack ?

Reply:
@marek-safar If in a struct there is a decimal field - the minimal alignment should be 8 bytes at least (on x64).
When Pack is set - this logic is ignored. e.g.:
```csharp
[StructLayout(LayoutKind.Sequential)]
public struct WithDecimal
{
    public decimal d; 
    public byte b;
} // Size: 24 (Mono used to return 20 before my fix)

[StructLayout(LayoutKind.Sequential, Pack = 1)]
public struct WithDecimal
{
    public decimal d;
    public byte b;
} // Size: 17 
```

Reply:
I don't know the story behind it in .NET Core, probably the last int field is used as long in some cases or some compatibility issues

Reply:
```
typedef struct tagDEC {
    USHORT wReserved;
    union {
        struct {
            BYTE scale;
            BYTE sign;
        };
        USHORT signscale;
    };
    ULONG Hi32;
    union {
        struct {
#ifdef _MAC
            ULONG Mid32;
            ULONG Lo32;
#else
            ULONG Lo32;
            ULONG Mid32;
#endif
        };
        ULONGLONG Lo64;
    };
} DECIMAL;
```